File: scripts/migrate_uploads_to_railway.mjs
/**
 * Dependency-free, resumable upload migration for the Railway volume.
 * Requires PITAYA_MIGRATION_ADMIN_TOKEN at runtime; the token is never saved.
 */

import { readdir, open, stat } from 'node:fs/promises';
import { join, relative, resolve, sep } from 'node:path';

const CHUNK_SIZE = 20 * 1024 * 1024;
const RETRY_COUNT = 5;
const baseUrl = (process.env.PITAYA_MIGRATION_URL || 'https://pitayafarm-production.up.railway.app').replace(/\/$/, '');
const adminToken = process.env.PITAYA_MIGRATION_ADMIN_TOKEN;
const uploadsRoot = resolve(process.env.PITAYA_UPLOADS_DIR || 'uploads');

if (!adminToken) throw new Error('PITAYA_MIGRATION_ADMIN_TOKEN is required.');

async function filesIn(directory) {
  const entries = await readdir(directory, { withFileTypes: true });
  const nested = await Promise.all(entries.map(async entry => {
    const path = join(directory, entry.name);
    return entry.isDirectory() ? filesIn(path) : entry.isFile() ? [path] : [];
  }));
  return nested.flat();
}

async function request(path, options = {}) {
  let lastError;
  for (let attempt = 0; attempt < RETRY_COUNT; attempt += 1) {
    try {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), 300000);
      const response = await fetch(`${baseUrl}${path}`, {
        ...options,
        headers: { Authorization: `Bearer ${adminToken}`, ...(options.headers || {}) },
        signal: controller.signal,
      });
      clearTimeout(timer);
      if (response.status < 500) return response;
      lastError = new Error(`Server returned ${response.status}: ${(await response.text()).slice(0, 300)}`);
    } catch (error) {
      lastError = error;
    }
    await new Promise(resolveWait => setTimeout(resolveWait, Math.min(30000, 2 ** attempt * 1000)));
  }
  throw lastError;
}

async function migrationStatus(path, size) {
  const response = await request('/api/admin/uploads/migration-status', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ path, size }),
  });
  const payload = await response.json();
  if (!response.ok || !payload.success) throw new Error(`Status failed for ${path}: ${payload.error || response.status}`);
  return payload;
}

async function migrateFile(path) {
  const relativePath = relative(uploadsRoot, path).split(sep).join('/');
  const { size } = await stat(path);
  let remote = await migrationStatus(relativePath, size);
  if (remote.complete) return { size, skipped: true };
  let offset = Number(remote.offset || 0);
  if (offset < 0 || offset > size) throw new Error(`Invalid resume offset for ${relativePath}: ${offset}`);

  const source = await open(path, 'r');
  try {
    while (offset < size) {
      const bytes = Math.min(CHUNK_SIZE, size - offset);
      const buffer = Buffer.allocUnsafe(bytes);
      const { bytesRead } = await source.read(buffer, 0, bytes, offset);
      if (!bytesRead) throw new Error(`Unable to read ${relativePath} at ${offset}`);

      const form = new FormData();
      form.append('path', relativePath);
      form.append('offset', String(offset));
      form.append('total_size', String(size));
      form.append('chunk', new Blob([buffer.subarray(0, bytesRead)], { type: 'application/octet-stream' }), 'chunk.bin');
      const response = await request('/api/admin/uploads/migrate-chunk', { method: 'POST', body: form });
      const payload = await response.json();
      if (response.status === 409) {
        offset = Number(payload.offset);
        continue;
      }
      if (!response.ok || !payload.success) throw new Error(`Upload failed for ${relativePath}: ${payload.error || response.status}`);
      const nextOffset = Number(payload.offset);
      if (nextOffset <= offset || nextOffset > size) throw new Error(`Invalid remote offset for ${relativePath}: ${nextOffset}`);
      offset = nextOffset;
    }
  } finally {
    await source.close();
  }

  remote = await migrationStatus(relativePath, size);
  if (!remote.complete || Number(remote.offset) !== size) throw new Error(`Incomplete migration for ${relativePath}`);
  return { size, skipped: false };
}

const imageExtensions = new Set(['.jpg', '.jpeg', '.png', '.webp', '.gif']);
const isImage = path => imageExtensions.has(path.slice(path.lastIndexOf('.')).toLowerCase());
const files = (await filesIn(uploadsRoot)).sort((left, right) => {
  const priorityDifference = Number(!isImage(left)) - Number(!isImage(right));
  return priorityDifference || left.localeCompare(right);
});
const totalBytes = (await Promise.all(files.map(file => stat(file)))).reduce((sum, item) => sum + item.size, 0);
console.log(`Migrating ${files.length.toLocaleString()} files (${(totalBytes / 1024 ** 3).toFixed(2)} GB). Restarting is safe.`);

let migratedBytes = 0;
let skippedFiles = 0;
for (let index = 0; index < files.length; index += 1) {
  const result = await migrateFile(files[index]);
  migratedBytes += result.size;
  skippedFiles += Number(result.skipped);
  console.log(`[${index + 1}/${files.length}] ${relative(uploadsRoot, files[index])} (${(migratedBytes / 1024 ** 3).toFixed(2)}/${(totalBytes / 1024 ** 3).toFixed(2)} GB)`);
}
console.log(`Complete. ${files.length.toLocaleString()} files verified; ${skippedFiles.toLocaleString()} already existed on Railway.`);

File: scripts/migrate_uploads_to_railway.py
"""Resumable transfer of local upload files to the Railway volume.

Run only after the matching server endpoints have been deployed.  The script
never stores credentials; supply the Railway site URL and admin token at run
time.  It asks the server for each file's current offset, so it can safely be
restarted after a network interruption.
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
import uuid
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen


CHUNK_SIZE = 20 * 1024 * 1024
RETRY_COUNT = 5


class HttpResponse:
    def __init__(self, status_code, content):
        self.status_code = status_code
        self.content = content
        self.text = content.decode("utf-8", errors="replace")

    @property
    def ok(self):
        return 200 <= self.status_code < 300

    def json(self):
        return json.loads(self.text)


class MigrationSession:
    """Small standard-library HTTP client so the migrator has no extra dependency."""

    def __init__(self, admin_token):
        self.headers = {"Authorization": f"Bearer {admin_token}"}

    def post(self, url, json=None, data=None, files=None, timeout=None):
        headers = dict(self.headers)
        if json is not None:
            payload = __import__("json").dumps(json).encode("utf-8")
            headers["Content-Type"] = "application/json"
        elif files:
            boundary = f"----pitaya-{uuid.uuid4().hex}"
            headers["Content-Type"] = f"multipart/form-data; boundary={boundary}"
            sections = []
            for name, value in (data or {}).items():
                sections.extend(
                    [
                        f"--{boundary}\r\n".encode(),
                        f'Content-Disposition: form-data; name="{name}"\r\n\r\n'.encode(),
                        str(value).encode(),
                        b"\r\n",
                    ]
                )
            for name, (filename, chunk, content_type) in files.items():
                sections.extend(
                    [
                        f"--{boundary}\r\n".encode(),
                        f'Content-Disposition: form-data; name="{name}"; filename="{filename}"\r\n'.encode(),
                        f"Content-Type: {content_type}\r\n\r\n".encode(),
                        chunk,
                        b"\r\n",
                    ]
                )
            sections.append(f"--{boundary}--\r\n".encode())
            payload = b"".join(sections)
        else:
            payload = b""

        request = Request(url, data=payload, headers=headers, method="POST")
        try:
            with urlopen(request, timeout=300) as response:
                return HttpResponse(response.status, response.read())
        except HTTPError as error:
            return HttpResponse(error.code, error.read())


def request_with_retry(method, url, **kwargs):
    last_error = None
    for attempt in range(RETRY_COUNT):
        try:
            response = method(url, timeout=(20, 300), **kwargs)
            if response.status_code < 500:
                return response
            last_error = RuntimeError(f"Server returned {response.status_code}: {response.text[:300]}")
        except (OSError, URLError) as error:
            last_error = error
        time.sleep(min(30, 2**attempt))
    raise RuntimeError(f"Request failed after {RETRY_COUNT} attempts: {last_error}")


def status(session, base_url, relative_path, size):
    response = request_with_retry(
        session.post,
        f"{base_url}/api/admin/uploads/migration-status",
        json={"path": relative_path, "size": size},
    )
    if not response.ok:
        raise RuntimeError(f"Status failed for {relative_path}: {response.status_code} {response.text}")
    payload = response.json()
    if not payload.get("success"):
        raise RuntimeError(f"Status failed for {relative_path}: {payload.get('error', 'unknown error')}")
    return bool(payload.get("complete")), int(payload.get("offset", 0))


def migrate_file(session, base_url, root, path):
    relative_path = path.relative_to(root).as_posix()
    total_size = path.stat().st_size
    complete, offset = status(session, base_url, relative_path, total_size)
    if complete:
        return total_size, True
    if offset < 0 or offset > total_size:
        raise RuntimeError(f"Invalid resume offset for {relative_path}: {offset}")

    with path.open("rb") as source:
        source.seek(offset)
        while offset < total_size:
            chunk = source.read(min(CHUNK_SIZE, total_size - offset))
            response = request_with_retry(
                session.post,
                f"{base_url}/api/admin/uploads/migrate-chunk",
                data={"path": relative_path, "offset": str(offset), "total_size": str(total_size)},
                files={"chunk": ("chunk.bin", chunk, "application/octet-stream")},
            )
            if response.status_code == 409:
                remote_offset = int(response.json().get("offset", -1))
                if remote_offset < 0 or remote_offset > total_size:
                    raise RuntimeError(f"Invalid conflict offset for {relative_path}: {response.text}")
                offset = remote_offset
                source.seek(offset)
                continue
            if not response.ok:
                raise RuntimeError(f"Upload failed for {relative_path}: {response.status_code} {response.text}")

            payload = response.json()
            if not payload.get("success"):
                raise RuntimeError(f"Upload failed for {relative_path}: {payload.get('error', 'unknown error')}")
            remote_offset = int(payload.get("offset", -1))
            if remote_offset <= offset or remote_offset > total_size:
                raise RuntimeError(f"Invalid upload offset for {relative_path}: {remote_offset}")
            offset = remote_offset
            source.seek(offset)

    complete, final_offset = status(session, base_url, relative_path, total_size)
    if not complete or final_offset != total_size:
        raise RuntimeError(f"Incomplete migration for {relative_path}")
    return total_size, False


def main():
    parser = argparse.ArgumentParser(description="Resumably migrate PITAYA uploads to Railway.")
    parser.add_argument("--url", required=True, help="Railway public app URL")
    parser.add_argument(
        "--admin-token",
        default=os.environ.get("PITAYA_MIGRATION_ADMIN_TOKEN"),
        help="Railway ADMIN_TOKEN (or set PITAYA_MIGRATION_ADMIN_TOKEN)",
    )
    parser.add_argument("--uploads", default="uploads", help="Local uploads folder")
    args = parser.parse_args()
    if not args.admin_token:
        parser.error("--admin-token or PITAYA_MIGRATION_ADMIN_TOKEN is required")

    root = Path(args.uploads).resolve()
    if not root.is_dir():
        raise SystemExit(f"Uploads folder does not exist: {root}")

    files = sorted(path for path in root.rglob("*") if path.is_file())
    total_bytes = sum(path.stat().st_size for path in files)
    print(f"Migrating {len(files):,} files ({total_bytes / 1024**3:.2f} GB). Restarting this command is safe.")

    session = MigrationSession(args.admin_token)
    base_url = args.url.rstrip("/")
    migrated_bytes = 0
    skipped_files = 0
    for index, path in enumerate(files, start=1):
        size, skipped = migrate_file(session, base_url, root, path)
        migrated_bytes += size
        skipped_files += int(skipped)
        print(f"[{index}/{len(files)}] {path.relative_to(root)} ({migrated_bytes / 1024**3:.2f}/{total_bytes / 1024**3:.2f} GB)")

    print(f"Complete. {len(files):,} files verified; {skipped_files:,} already existed on Railway.")


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\nPaused safely. Re-run the same command to resume.", file=sys.stderr)
        raise SystemExit(130)
